drivers/flash/flash_hp_ra.c

```diff
@@ -0,0 +1,392 @@
+{
+	int err = -ENOTSUP;
+
+	switch (code) {
+#if defined(CONFIG_FLASH_RA_WRITE_PROTECT)
+	case FLASH_RA_EX_OP_WRITE_PROTECT:
+		err = flash_ra_ex_op_write_protect(dev, in, out);
+		break;
+#endif /* CONFIG_FLASH_RA_WRITE_PROTECT */
+
+	default:
+		break;
+	}
+
+	return err;
+}
+#endif
+
+static int flash_ra_init(const struct device *dev)
+{
+	const struct device *dev_ctrl = DEVICE_DT_INST_GET(0);
+	const struct flash_hp_ra_config *cfg = dev->config;
+	struct flash_hp_ra_data *flash_data = dev->data;
+
+	if (!device_is_ready(dev_ctrl)) {
+		return -ENODEV;
+	}
+
+	if (cfg->flash_ra_parameters.write_block_size == UNIT_CODE_FLASH_PROGRAMMING) {
+		flash_data->FlashRegion = CODE_FLASH;
+	} else {
+		flash_data->FlashRegion = DATA_FLASH;
+	}
+
+	flash_data->controller = dev_ctrl->data;
+
+	return 0;
+}
+
+static void flash_controller_ra_irq_config_func(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	R_ICU->IELSR[DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, irq)] = ELC_EVENT_FCU_FRDYI;
+	R_ICU->IELSR[DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, irq)] = ELC_EVENT_FCU_FIFERR;
+
+	IRQ_CONNECT(DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, irq),
+		    DT_IRQ_BY_NAME(DT_DRV_INST(0), frdyi, priority), fcu_frdyi_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+	IRQ_CONNECT(DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, irq),
+		    DT_IRQ_BY_NAME(DT_DRV_INST(0), fiferr, priority), fcu_fiferr_isr,
+		    DEVICE_DT_INST_GET(0), 0);
+
+	irq_enable(DT_INST_IRQ_BY_NAME(0, frdyi, irq));
+	irq_enable(DT_INST_IRQ_BY_NAME(0, fiferr, irq));
+}
+
+static int flash_controller_ra_init(const struct device *dev)
+{
+	fsp_err_t err = FSP_SUCCESS;
+	const struct flash_hp_ra_controller_config *cfg = dev->config;
+	struct flash_hp_ra_controller *data = dev->data;
+
+	cfg->irq_config(dev);
+
+	err = R_FLASH_HP_Open(&data->flash_ctrl, &data->fsp_config);
+	LOG_DBG("flash: open error=%d", err);
+
+	if (err != FSP_SUCCESS) {
+		return -EIO;
+	}
+
+	k_sem_init(&data->ctrl_sem, 1, 1);
+
+	return 0;
+}
+
+static struct flash_hp_ra_controller_config flash_hp_ra_controller_config = {
+	.irq_config = flash_controller_ra_irq_config_func,
+};
+
+#define RA_FLASH_INIT(index)                                                                       \
+	static struct flash_hp_ra_config flash_hp_ra_config_##index = {                            \
+		.flash_ra_parameters = {                                                           \
+			.write_block_size = DT_PROP(index, write_block_size),                      \
+			.erase_value = 0xff,                                                       \
+		}};                                                                                \
+	struct flash_hp_ra_data flash_hp_ra_data_##index = {                                       \
+		.area_address = DT_REG_ADDR(index),                                                \
+		.area_size = DT_REG_SIZE(index),                                                   \
+		.erase_block_size = DT_PROP(index, erase_block_size)};                             \
```

Comment:
No check against wbs limitations of hardware.

Reply:
I do not use the `erase_block_size` anymore because the purpose of this is I intend to calculate the page_count and page_size for data region. But as I mentioned before, the HAL is already defined for this size so I would like to use it instead of define here. 
Thanks very much.

Reply:
Yeah, but note that users may overlay your DTS with what they want which means that you can either check that it has been what you expect or wait for support case with questions.

Reply:
Got it. I added the handle to check the size of erase_block_size based on the flash region in `flash_ra_init`. The size must be multiple of the hardware specifies.
Thanks very much.

Reply:
OK. Btw, If you look at some other drivers you may find how they do the check to fail compilation; this means that there is no run-time code and you know at compile time that DTS gone bad.
I am acking the PR as is, it is up to you if you want to have compile time or run-time check.

Reply:
I see. I updated the erase_block_size to have compiler time check for both regions, no matter what region the users want to program, it must pass the condition. Thanks again for letting me know.